loadgen/generic_executor.go

```diff
@@ -68,34 +71,52 @@ func (g *GenericExecutor) newRun(info ScenarioInfo) (*genericRun, error) {
 // iterations is reached.
 func (g *genericRun) Run(ctx context.Context) error {
 	ctx, cancel := context.WithCancel(ctx)
-	if g.config.Duration > 0 {
-		ctx, cancel = context.WithTimeout(ctx, g.config.Duration)
+	if g.config.Timeout > 0 {
+		g.logger.Debugf("Will timeout after %v", g.config.Timeout)
+		ctx, cancel = context.WithTimeout(ctx, g.config.Timeout)
 	}
 	defer cancel()
 
+	durationCtx := ctx
+	if g.config.Duration > 0 {
+		durationCtx, cancel = context.WithTimeout(ctx, g.config.Duration)
+		defer cancel()
+	}
+
 	startTime := time.Now()
 	var runErr error
 	doneCh := make(chan error)
 	var currentlyRunning int
-	waitOne := func() {
+	waitOne := func(exitOnDurationDone bool) {
+		timeoutOrPending := make(<-chan struct{})
```

Comment:
Just an FYI, this also works I believe (no need to change anything)

```suggestion
		var timeoutOrPending <-chan struct{}
```